### PR TITLE
Octree

### PR DESCRIPTION
**Pull request by ttnghia**
Here is our new octree example 😈 
@mosra, feel free to edit it. I could not think of a nicer teaser image. The simulation scene can be changed, such as the number of spheres, sphere size, sphere color etc.

octree

### CODE REVIEW

File: src/octree/LooseOctree.h [hunk 1]
```diff
@@ -0,0 +1,331 @@
...
+    const Vector3 _center;        /* center of the tree */
+    const Float   _halfWidth;     /* width of the tree bounding box */
+
+    const Float _minHalfWidth;    /* minimum width allowed for the tree nodes */
+    std::size_t _maxDepth;        /* max depth of the tree, which is computed based on _minWidth */
+
+    OctreeNode* const _pRootNode; /* root node, should not be reassigned throughout the existence of the tree */
+
+    /* Store the free node blocks (8 nodes) that can be used right away */
+    std::vector<OctreeNodeBlock*> _freeNodeBlocks;
```
[the remark this answers is not in the document]
Comment (ttnghia) [reply src/octree/LooseOctree.h:315]: The vector content is just pointers, thus they can be copy/move around without changing their memory contents.

File: src/octree/LooseOctree.cpp [hunk 2]
```diff
@@ -0,0 +1,356 @@
...
+        for(std::size_t childIdx = 0; childIdx < 8; ++childIdx) {
+            auto newCenter = _center;
+            newCenter[0] += (childIdx & 1) ? childHalfWidth : -childHalfWidth;
+            newCenter[1] += (childIdx & 2) ? childHalfWidth : -childHalfWidth;
+            newCenter[2] += (childIdx & 4) ? childHalfWidth : -childHalfWidth;
+
+            OctreeNode* const pChildNode = &_pChildren->_nodes[childIdx];
+
+            /* Placement new: re-use existing memory block, just call constructor to re-initialize data */
+            new(pChildNode) OctreeNode(_pTree, this, newCenter, childHalfWidth, _depth + 1u);
```
[the remark this answers is not in the document]
Comment (ttnghia) [reply src/octree/LooseOctree.cpp:103]: Sure.

File: src/octree/LooseOctree.h [hunk 3]
```diff
@@ -0,0 +1,331 @@
...
+     */
+    bool contains(const Vector3& point) const { return contains(point[0], point[1], point[2]); }
+    bool contains(const Float x, const Float y, const Float z) const {
+        return x > _lowerBound[0]
+               && y > _lowerBound[1]
+               && z > _lowerBound[2]
+               && x < _upperBound[0]
+               && y < _upperBound[1]
+               && z < _upperBound[2];
+    }
```
[the remark this answers is not in the document]
Comment (ttnghia) [reply src/octree/LooseOctree.h:142]: If a point is exactly at the bounds of a node, then that point can be stored either at that node or at one of the neighbor nodes. In both cases, that point will always be identified from the parent node.

File: src/octree/LooseOctree.h [hunk 4]
```diff
@@ -0,0 +1,331 @@
...
+    /* Constructor called during node splitting */
+    explicit OctreeNode(LooseOctree* const tree, OctreeNode* const pParent,
+                        const Vector3& nodeCenter, const Float halfWidth,
+                        const std::size_t depth);
+
+    /* Common properties */
+    bool isLeaf() const { return _bIsLeaf; }
+    const Vector3 getCenter() const { return _center; }
+    Float getHalfWidth() const { return _halfWidth; }
+    std::size_t getDepth() const { return _depth; }
```
[the remark this answers is not in the document]
Comment (ttnghia) [reply src/octree/LooseOctree.h:93]: Ok.

File: src/octree/OctreeExample.cpp [hunk 5]
```diff
@@ -0,0 +1,540 @@
...
+    if(event.modifiers() & MouseMoveEvent::Modifier::Shift) {
+        _arcballCamera->translate(event.position());
+    } else { _arcballCamera->rotate(event.position()); }
+    event.setAccepted();
+    redraw(); /* camera has changed, redraw! */
+}
+
+void OctreeExample::mouseScrollEvent(MouseScrollEvent& event) {
+    const Float delta = event.offset().y();
+    if(Math::abs(delta) < 1.0e-2f) { return; }
```
[the remark this answers is not in the document]
Comment (ttnghia) [reply src/octree/OctreeExample.cpp:533]: My automatic formatter refuses to do that 😄

File: src/octree/OctreeExample.cpp [hunk 6]
```diff
@@ -0,0 +1,540 @@
...
+                                                              Color3(0.1f, 0.5f, 0.6f),
+                                                              Matrix4::scaling(Vector3{ 1 }),
+                                                              *_drawables.get() }
+                        );
+                }
+                _drawableBoxes[boxIdx]->setEnable(true);
+                _boxObjs[boxIdx++]->setTransformation(
+                    Matrix4::translation(pNode->getCenter()) *
+                    Matrix4::scaling(Vector3{ pNode->getHalfWidth() })
+                    );
```
[the remark this answers is not in the document]
Comment (ttnghia) [reply src/octree/OctreeExample.cpp:466]: Yeah, why I didn't think of this?

File: src/octree/LooseOctree.cpp [hunk 7]
```diff
@@ -0,0 +1,356 @@
...
+void LooseOctree::setPoints(std::vector<Vector3>& points) {
+    clearPoints();
+    const std::size_t nPoints = points.size();
+    if(nPoints == 0) {
+        return;
+    }
+    _octreePoints.resize(nPoints);
+    for(std::size_t idx = 0; idx < nPoints; ++idx) {
+        const auto pPoint = &_octreePoints[idx];
+        new(pPoint) OctreePoint(&points, idx); /* placement new */
```
[the remark this answers is not in the document]
Comment (ttnghia) [reply src/octree/LooseOctree.cpp:217]: I change this to a plain assignment.
OctreeNode cannot be re-assign, as the class has constant variable.

File: src/octree/OctreeExample.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (ttnghia) [reply src/octree/OctreeExample.cpp:454]: So you have to do that on your side :)

**Comment by ttnghia** [comment]
@mosra The code looks much better now 😃

**Comment by ttnghia** [comment]
I am curious to see this works in WebGL 😄

**Comment by mosra** [comment]
Thank you! The last unresolved comment is https://github.com/mosra/magnum-examples/pull/86/files#r441446247, about making it possible to switch and benchmark the bruteforce method at runtime -- just want to know your opinion on that, and I'm fine with doing that myself post-merge :)

File: src/octree/OctreeExample.cpp [hunk 8]
```diff
@@ -0,0 +1,540 @@
...
+                                                  Matrix4::scaling(Vector3{ 1 }),
+                                                  *_drawables.get() }
+            );
+
+        /* Draw the remaining nodes */
+        updateTreeNodeBoundingBoxes();
+    }
+
+    /* Run benchmark */
+    if(args.value<std::size_t>("benchmark")) {
```
[the remark this answers is not in the document]
Comment (ttnghia) [reply src/octree/OctreeExample.cpp:240]: Done.

**Comment by ttnghia** [comment]
Done.

**Comment by mosra** [comment]
Merged in 5ff5175d9d5d820374108aa52355a7b65d0cf967, thank you! :)

**Pull request closed without merging**